staging/src/k8s.io/apiserver/plugin/pkg/kmsprovider/vault/vault.go

```diff
@@ -0,0 +1,218 @@
+	configFileContents, err := ioutil.ReadAll(configFile)
+	if err != nil {
+		return nil, fmt.Errorf("could not read contents: %v", err)
+	}
+
+	var config EnvelopeConfig
+	err = yaml.Unmarshal(configFileContents, &config)
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing file: %v", err)
+	}
+
+	err = checkConfig(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := newClientWrapper(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &vaultEnvelopeService{config: &config, client: client}, nil
+}
+
+func checkConfig(config *EnvelopeConfig) error {
+	if len(config.KeyNames) == 0 {
+		return errors.New("vault provider has no valid key names")
+	}
+
+	if config.Address == "" {
+		return errors.New("vault provider has no valid address")
+	}
+
+	return checkAuthConfig(config)
+}
+
+func checkAuthConfig(config *EnvelopeConfig) error {
+	var count uint
+
+	if config.Token != "" {
+		count++
+	}
+
+	if config.ClientCert != "" || config.ClientKey != "" {
+		if config.ClientCert == "" || config.ClientKey == "" {
+			return errors.New("vault provider has invalid TLS authentication information")
+		}
+		count++
+	}
+
+	if config.RoleID != "" || config.SecretID != "" {
+		if config.RoleID == "" {
+			return errors.New("vault provider has invalid approle authentication information")
+		}
+		count++
+	}
+
+	if count == 0 {
+		return errors.New("vault provider has no authentication information")
+	}
+	if count > 1 {
+		return errors.New("vault provider has more than one authentication information")
+	}
+
+	return nil
+}
+
+type vaultEnvelopeService struct {
+	config *EnvelopeConfig
+	client *clientWrapper
+	// We may update token for api.Client, but there is no sync for api.Client.
+	// Read lock for encrypt/decrypt requests, write lock for login requests which
+	// will update token for api.Client.
+	rwmutex sync.RWMutex
+}
+
+func (s *vaultEnvelopeService) Decrypt(data string) ([]byte, error) {
+	// Find the mached key
+	var key string
+	for _, name := range s.config.KeyNames {
+		if strings.HasPrefix(data, name+":") {
+			key = name
+			break
+		}
+	}
+	if key == "" {
+		return nil, errors.New("no matching vault key found")
+	}
+
+	// Replace the key name with "vault:" for Vault transit API
+	cipher := strings.Replace(data, key, vaultPrefix, 1)
+
+	//plain, _, err := s.client.decrypt(key, cipher)
+	plain, err := s.withRefreshToken((*clientWrapper).decrypt, key, cipher)
+	if err != nil {
+		return nil, err
+	}
+
+	return base64.StdEncoding.DecodeString(plain)
+}
+
+func (s *vaultEnvelopeService) Encrypt(data []byte) (string, error) {
+	// Use the frist key to encrypt
+	key := s.config.KeyNames[0]
+	plain := base64.StdEncoding.EncodeToString(data)
+
+	cipher, err := s.withRefreshToken((*clientWrapper).encrypt, key, plain)
+	if err != nil {
+		return "", err
+	}
+
+	// The format of cipher from Vault is "vault:v1:....".
+	// "vault:" is unnecessary, replace it with key name.
```

Comment:
can I get a prepend instead?  Just something about it rubs me the wrong way.

Reply:
@deads2k I did not get your comment, can you please elaborate?

Reply:
instead of `strings.Replace`, could it be `<key>:vault:<remainder>`?

Reply:
ok, I have removed the use of strings.Replace() and used hasPrefix() as per @ericchiang comment.